examples/server/utils.hpp

```diff
@@ -913,3 +960,335 @@ static std::vector<common_adapter_lora_info> parse_lora_request(
 
     return lora;
 }
+
+//
+// utils for interacting with libmtmd
+// (may need to refactor in near future)
+//
+
+// each chunk can contain either one SINGLE text token or pointer to image
+// this is to simplify the logic of KV cache management
+struct server_token {
+    llama_token txt;
+    std::shared_ptr<mtmd_image_tokens> img;
+    std::string str() const {
+        // for debugging
+        GGML_ASSERT(img || txt != LLAMA_TOKEN_NULL);
+        if (img) {
+            return "<embd> ";
+        } else {
+            return std::to_string(txt) + " ";
+        }
+    }
+};
+
+/**
+ * server_tokens is a helper to manage the input tokens and image for the server.
+ * it is made this way to simplify the logic of KV cache management.
+ *
+ * each token can be either a text token or a pointer to an image.
+ * if image usually contains multiple tokens, each token contains a shared_ptr to the same image.
+ */
+struct server_tokens {
+    bool has_mtmd = false;
+    std::vector<server_token> values;
+
```

Comment:
I think you can maintain `llama_tokens values_text;` in parallel:

```c++
struct server_tokens {
    bool has_mtmd = false;
    std::vector<server_token> values;
    llama_tokens values_text; // <-- keep in sync with values

...
```

This way you would avoid a lot of copies and memory allocations when converting from `server_tokens` to `llama_tokens`.

Reply:
Hmm your suggestion is a bit like what I'm currently doing in https://github.com/ngxson/wllama/pull/173 , what I did is to use all negative numbers from `-1000` (`count--` each time) to mark the image placement.

Not sure if this looks right to you, but another option could be to simply use `LLAMA_TOKEN_NULL` to mark the image token, then have a mapping `llama_pos --> image` to keep track of the image placement. This way, we don't need the `count--` hack that I mentioned above.

Reply:
This also works functionally, though you would still need to have the `llama_tokens server_tokens::get_text_tokens()` method.

I think ideally you want to avoid this method so that `server_tokens.values_text` could be used interchangeably with the legacy `llama_tokens`. It will not have image tokens or placeholders - just the text tokens. The goal is when multi-modality is not used, to have the same logic in the code as before and also to avoid copies of the tokens on each iteration (for example during speculative decoding).

Reply:
> Not sure if this looks right to you, but another option could be to simply use LLAMA_TOKEN_NULL to mark the image token, then have a mapping llama_pos --> image to keep track of the image placement. This way, we don't need the count-- hack that I mentioned above.

Yes, actually this should work too and you don't need `get_text_tokens()`. Maybe this is the better option.